### PR TITLE
Handling for multiple dictionary files and selection of the latest version

### PR DESCRIPTION
**Pull request by rvyuha**
Created a new utility function found at version-utils.R.
The function is responsible for parsing a single string or vector of string with file names and returning the latest version as well as file name with the latest version.
Modified _dictionary-version.qmd to utilize the new function.

### CODE REVIEW

**Comment by rvyuha** [comment]
Warning contains extra dictionaries to allow for testing. Should remove other versions prior to merge.

File: R/version-utils.R [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+  for (file_name in file_names) {
+    version_number <-
+      regmatches(
+        file_name,
+        regexec(dictionary_version_pattern, file_name)
+      )[[1]][2]
+    version_numbers <- append(version_numbers, version_number)
+  }
+  
+  formated_versions <- semver::parse_version(version_numbers)
```
Comment (yulric) [inline R/version-utils.R:26]: Rename to `parsed_version`?

File: qmd/_dictionary-version.qmd [hunk 2]
```diff
@@ -1,9 +1,14 @@
 <!--Source constants file until I get clarity for environment usage across entire package or if sourcing is enough -->
 `r source("R/constants.R")`
 ```{r, echo=FALSE}
-dictionary_full_file_name <- list.files(file.path(getwd(),constants$dictionary_directory), pattern = "^ODM.*xlsx$")
-dictionary_version_pattern <- "ODM_dictionary_(.*?).xlsx"
+source("R/version-utils.R")
+
+dictionary_full_file_names <- list.files(file.path(getwd(),constants$dictionary_directory), pattern = "^ODM.*xlsx$")
```
Comment (yulric) [inline qmd/_dictionary-version.qmd:6]: Why are we using two patterns? This one and the one in the `get_latest_version` function?

**Review by yulric — CHANGES_REQUESTED** [review]

File: R/version-utils.R [hunk 3]
```diff
@@ -0,0 +1,40 @@
+#' Get latest version
+#' 
+#' Get the latest version from vector of file names
+#' 
+#' @param file_names string or string vector containing file names with valid semantic versioning
+#' 
+#' @return vector of strings containing latest version and latest version file name 
+get_latest_version <- function(file_names) {
+  # Display warning for multiple dictionaries as only 1 should be stored on github
+  if(length(file_names)>1){
```
[the remark this answers is not in the document]
Comment (rvyuha) [reply R/version-utils.R:10]: It would not trigger since a single string is less then 1 in length. Despite the number of characters in a string.

**Review by yulric — CHANGES_REQUESTED** [review]
1. Don't think this [comment](https://github.com/Big-Life-Lab/PHES-ODM-Doc/pull/32#discussion_r1138704699) was addressed. To be more specific, I was proposing the creation of a function to get the latest dictionary file from the file system and its version.As opposed to what you did here where you're splitting up the retreival of different dictionary files and selecting the latest one.
2. Don't think this [comment](https://github.com/Big-Life-Lab/PHES-ODM-Doc/pull/32#discussion_r1138707396) was addressed

**Comment by rvyuha** [comment]
@yulric  Ready for review

File: R/version-utils.R [hunk 4]
```diff
@@ -0,0 +1,41 @@
+source("R/constants.R")
+source("R/semver-handling.R")
+#' Get latest version
+#'
+#' Get the latest version from vector of file names
+#'
+#' @param file_names string or string vector containing file names with valid semantic versioning
+#'
+#' @return vector of strings containing latest version and latest version file name
+get_latest_version <- function() {
```
Comment (yulric) [inline R/version-utils.R:10]: Can we change this function to return the dictionary file to use and version, rather then the name of the dictionary file name and version number?

It looks like the parts, sets, and tables qmds don't need the dictionary file name for any reason other than to get the dictionary itself.

Comment (yulric) [reply R/version-utils.R:10]: get_latest_version -> get_dictionary_file

File: R/semver-handling.R [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+#' Convert to semver
+#' 
+#' Utility function to break apart a version string into a version list.
+#' With further conversion of alphanumeric pre-release info into numeric list.
+#' 
+#' @param version_string string representing the version as one full string.
+#' 
+#' @return list containing the version split into seperate elements.
+convert_to_semver <- function(version_string) {
+  version_extraction_regex <-
```
Comment (yulric) [inline R/semver-handling.R:58]: Rename to `semver_regex`

Comment (yulric) [reply R/semver-handling.R:58]: Can you add a comment saying where you got the regex from

File: R/semver-handling.R [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+    '(0|[1-9]\\d*)\\.(0|[1-9]\\d*)\\.(0|[1-9]\\d*)(?:-((?:0|[1-9]\\d*|\\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\\.(?:0|[1-9]\\d*|\\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\\+([0-9a-zA-Z-]+(?:\\.[0-9a-zA-Z-]+)*))?'
+  
+  semantic_versions <- list("alpha", "beta", "rc")
```
Comment (yulric) [inline R/semver-handling.R:61]: Rename to `supported_pre_releases`

File: R/semver-handling.R [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+  # Elements 2-4 are the Major, Minor, and Path digits
+  # Element 5 is <pre-release> information according to https://semver.org/#backusnaur-form-grammar-for-valid-semver-versions
+  # Element 6 is <build> information according to https://semver.org/#backusnaur-form-grammar-for-valid-semver-versions
+  formated_version <- regmatches(version_string,
+                                 regexec(version_extraction_regex, version_string))[[1]][1:5]
+  
+  # Split semantic pre-release info
+  pre_release <- formated_version[[5]]
+  if(pre_release != ""){
+  split_release_info <- strsplit(pre_release,"\\.")
```
Comment (yulric) [inline R/semver-handling.R:74]: Can you indent the lines so that they're within the if loop?

Comment (yulric) [reply R/semver-handling.R:74]: Rename to `pre_release_info`

File: R/semver-handling.R [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+  semantic_relase <- split_release_info[[1]][[1]]
```
Comment (yulric) [inline R/semver-handling.R:75]: `pre_relase_name`

File: R/semver-handling.R [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+    for (semantic_weight in seq_len(length(semantic_versions))) {
```
Comment (yulric) [inline R/semver-handling.R:76]: semantic_weight -> supported_pre_release

File: R/semver-handling.R (hunk 6 above)
Comment (yulric) [reply R/semver-handling.R:61]: Why aren't you using a vector here?

File: R/semver-handling.R [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+
+#' Convert to semver
+#' 
+#' Utility function to break apart a version string into a version list.
+#' With further conversion of alphanumeric pre-release info into numeric list.
+#' 
+#' @param version_string string representing the version as one full string.
+#' 
+#' @return list containing the version split into seperate elements.
+convert_to_semver <- function(version_string) {
```
Comment (yulric) [inline R/semver-handling.R:57]: I think this function should only be responsible for parsing the version information out of the version string. It should not do anything related to assigning a weight to the pre-release information. 

It should parse the semver string and return a named list that looks like:

list(
    major = 1,
    minor = 2,
    patch = 3,
    pre_release = "alpha",
    pre_release_version = 1
)

All the work for assigning a weigh can be done in a sorting, comparison function

File: R/semver-handling.R [hunk 11]
```diff
@@ -0,0 +1,93 @@
+#' Get max version
+#' 
+#' Extract max version from vector of version strings.
+#' 
+#' @param version_strings vector of string containing version string.
+#' 
+#' @return string representing the max version found
+get_max_version <- function(versions_strings) {
+  versions_list <- list()
```
Comment (yulric) [inline R/semver-handling.R:9]: Why are you using a list as opposed to a vector?

File: R/semver-handling.R [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+  for (version_index in seq_len(length(versions_strings))) {
+    versions_list[[version_index]] <-
+      convert_to_semver(versions_strings[[version_index]])
+  }
+  # R does not appear to contain a do while and this is the suggested approach
+  version_element_index <- 2
+  max_version <- list()
```
Comment (yulric) [inline R/semver-handling.R:16]: Why are you using a list as opposed to a vector?

Comment (yulric) [reply R/semver-handling.R:16]: max_version -> sorted_versions

File: R/semver-handling.R [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+  repeat {
```
Comment (yulric) [inline R/semver-handling.R:17]: I'm confused how this looping is working, can you make a flow diagram and send it to me

**Review by yulric — CHANGES_REQUESTED** [review]
1. Rename version-utils.R to the function within it

File: R/semver-handling.R (hunk 11 above)
Comment (rvyuha) [reply R/semver-handling.R:9]: A vector can only store one dimensional same type objects. This stores list objects returned by my semver function.

File: R/semver-handling.R (hunk 12 above)
Comment (rvyuha) [reply R/semver-handling.R:16]: > Why are you using a list as opposed to a vector?

A vector is unable to store lists.

File: R/semver-handling.R (hunk 9 above)
Comment (yulric) [reply R/semver-handling.R]: Was mistaken what the `semantic_weight` variable contains, better named as `supported_pre_release_index`

File: R/semver-handling.R (hunk 10 above)
Comment (yulric) [reply R/semver-handling.R:60]: Lets also throw an error in the following situations:

1. If there is more than one pre-release tag
2. If there is an un-supported pre-release tag
3. If the version of a pre-release tag cannot be converted to a semver itself

**Comment by rvyuha** [comment]
@yulric Ready for review

**Pull request merged**